### PR TITLE
Correlations: Migrate config type to root

### LINKED ISSUE #91643: Correlations: Migrate Config type to a database column
In correlations, the configuration is saved as JSON in the database, with the entirety of the JSON being in one column.

For #91535, we realized the config type would be relevant to validation. If the config type is external, there is no target datasource. 

Parsing the JSON to extract a value in a query is not great, as we support several different kinds of databases for Grafana and thus like to keep the queries simple, so they can stay within core querying functionalities in the different dialects that exist to create queries. However, the concept of adding a way to do limited JSON parsing was explored here https://github.com/grafana/grafana/pull/91535/commits/3466355c4db506eb91820cea4013c1cf6320aa2d 

- Remove Type from the correlation config json, and make it its own column
- For existing correlations, write a migration script that will extract values with Go and move them to the new column
- Ensure all correlation behavior, like saving and creating, work without regressions. Logic should fall back to looking for a type field inside of config, if the correlation type outside of config does not exist.
- In the database logic for correlations, add a WHERE clause that says the target datasource may be null if the config type is external

### PR DESCRIPTION
**Pull request by gelicia**
**What is this feature?**

In doing #89844, we recognized the need to do filtering at the SQL level to support a new scenario - when the correlation type is `external`, there will not be a target datasource. This is a valid correlation that should display.

The target datasource when the correlation type is `query` can be missing for various reasons that make the correlation invalid - such as the user not having permissions to the target datasource, or the target datasource no longer existing. In that case, the correlation should not display.

The correlation type is where the logic branches. However, we put this value in the `config` column of correlations, where it is stored with other values in JSON. It is not a simple matter to extract it while keeping support for various database dialects, so we are deprecating the value from the JSON and moving it to its own column, which will be in the main Correlations object.

The `external` type will be future work, so we are not putting anything in this PR for that. The only valid type is `query`, which is why we can just assume at this point that if the type is not defined in the root, it is `query`.

**Which issue(s) does this PR fix?**:

Fixes #91643 

**Special notes for your reviewer:**

Tested by creating a correlation in both editors and verifying it works. Cleared out the correlations table and created new from provisioning works as well.

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/contribute/release-notes/#how-to-determine-if-content-belongs-in-whats-new), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/contribute/release-notes/) doc.

### CODE REVIEW

File: pkg/services/sqlstore/migrations/correlations_mig.go [hunk 1]
```diff
@@ -63,4 +62,8 @@ func addCorrelationsMigrations(mg *Migrator) {
 	mg.AddMigration("add provisioning column", NewAddColumnMigration(correlationsV2, &Column{
 		Name: "provisioned", Type: DB_Bool, Nullable: false, Default: "0",
 	}))
+
+	mg.AddMigration("add type column", NewAddColumnMigration(correlationsV2, &Column{
```
Comment (aocenas) [inline pkg/services/sqlstore/migrations/correlations_mig.go:66]: Wouldn't it make sense to make this `nullable: false` and migrate now with default value of `query`? That way we prevent any accidental `null` values in the future and we don't have to do anything special on read path.

**Review by harisrozajac — APPROVED** [review]
Looks great. Tested this by creating and editing correlations and didn't find any regressions 🚀

File: pkg/services/sqlstore/migrations/correlations_mig.go (hunk 1 above)
Comment (aocenas) [reply pkg/services/sqlstore/migrations/correlations_mig.go:66]: So I don't think this kind of migration should be anything special, we already have the same non nullable with default columns in other places like here https://github.com/grafana/grafana/blob/d8904f3ca46cc8058036e5d01eba39f759fd581a/pkg/services/sqlstore/migrations/datasource_mig.go#L119

So yeah I can see it is failing but there seems to be something else going on which I think we should investigate a bit.

Comment (aocenas) [reply pkg/services/sqlstore/migrations/correlations_mig.go:66]:
```
migrations_mock.go:43: failed to initialize sql store: migration failed (id = add type column): Error 1064 (42000): You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near 'query' at line 1
```

The addColumn SQL statement that the migrator code creates does not seem to be very dialect specific so I wonder if the problem is `query` string itself, maybe it is a reserved word that needs to be quoted in some special way.

Comment (aocenas) [reply pkg/services/sqlstore/migrations/correlations_mig.go:66]: Would helpful to enable debug logging there for this as this log line https://github.com/grafana/grafana/blob/5dfe4cf407d0ae27249435693d97ac2710262267/pkg/services/sqlstore/migrator/migrator.go#L280 would be helpful.

Comment (gelicia) [reply pkg/services/sqlstore/migrations/correlations_mig.go:66]: This was fixed. A lot easier than I was thinking, just needed to look at other similar migrations that have text.

**Pull request merged**